src/coreclr/src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComponentActivator.cs

```diff
@@ -50,45 +75,19 @@ private static string MarshalToString(IntPtr arg, string argName)
-                                                            IntPtr typeNameNative,
-                                                            IntPtr methodNameNative,
-                                                            IntPtr delegateTypeNative,
-                                                            IntPtr reserved,
-                                                            IntPtr functionHandle)
+                                                                   IntPtr typeNameNative,
+                                                                   IntPtr methodNameNative,
+                                                                   IntPtr delegateTypeNative,
+                                                                   IntPtr reserved,
+                                                                   IntPtr functionHandle)
         {
             try
             {
-                // Load the assembly and create a resolver callback for types.
+                // Validate all parameters first.
                 string assemblyPath = MarshalToString(assemblyPathNative, nameof(assemblyPathNative));
-                IsolatedComponentLoadContext alc = GetIsolatedComponentLoadContext(assemblyPath);
-                Func<AssemblyName, Assembly> resolver = name => alc.LoadFromAssemblyName(name);
-
-                // Get the requested type.
                 string typeName = MarshalToString(typeNameNative, nameof(typeNameNative));
-                Type type = Type.GetType(typeName, resolver, null, throwOnError: true)!;
-
-                // Get the method name on the type.
                 string methodName = MarshalToString(methodNameNative, nameof(methodNameNative));
-
-                // Determine the signature of the type. There are 3 possibilities:
-                //  * No delegate type was supplied - use the default (i.e. ComponentEntryPoint).
-                //  * A sentinel value was supplied - the function is marked UnmanagedCallersOnly. This means
-                //      a function pointer can be returned without creating a delegate.
-                //  * A delegate type was supplied - Load the type and create a delegate for that method.
-                Type? delegateType;
-                if (delegateTypeNative == IntPtr.Zero)
-                {
-                    delegateType = typeof(ComponentEntryPoint);
-                }
-                else if (delegateTypeNative == (IntPtr)(-1))
-                {
-                    delegateType = null;
-                }
-                else
-                {
-                    string delegateTypeName = MarshalToString(delegateTypeNative, nameof(delegateTypeNative));
-                    delegateType = Type.GetType(delegateTypeName, resolver, null, throwOnError: true)!;
-                }
+                (Type? delegateType, string? delegateTypeName) = GetDelegateTypeArgs(delegateTypeNative);
```

Comment:
I think the logic is harder to follow now that this is split over 3 different functions. Was it really necessary to split it?

Reply:
Yes. https://github.com/dotnet/runtime/pull/37696#discussion_r438061553

Reply:
Can `InternalGetFunctionPointer` just take `delegateTypeNative`? 

I do not see the point by doing little bit of upfront work in `GetDelegateTypeArgs`, passing the tuple around, and finishing it in `InternalGetFunctionPointer`.

Reply:
The idea is to validate the parameters before doing anything meaningful like creating an ALC. Can't the Marshal call throw an exception if given a bad pointer?

Reply:
> Can't the Marshal call throw an exception if given a bad pointer?

Yes, but it's not really a parameter validation - it will try to read from the pointer, if that throws an AV it will turn into an exception.

I understand the goal here (to validate everything up-front), so I'm basically fine with this, but if @jkotas or others would prefer to not use this, I would also be OK if we duplicate the parameter validation (basically inline the method here as it was before, and then make a copy into the second one we're introducing) - or I would not mind if we simply postpone the marshal call to the point where we're going to actually get the delegate type.